### PR TITLE
Precompute ASR model features

### PR DESCRIPTION
**Pull request by cyaaronk**
I have implemented a python script that can precompute and save ESPNet2 ASR model features to disk.
For your preview. I will update the code and push a second version if it looks okay.

Features implemented:
1. batch processing
2. Automatic unpadding
3. String argument to specify module output to save. E.g. --output_modules=[encoder.encoders.*]
4. Use pytorch module forward hooks

Features not implemented yet:
1. Only encoder features can be saved. Haven't implemented for WavLM features yet.

### CODE REVIEW

**Comment by cyaaronk** [comment]
Updated code.

Features implemented:

    1. batch processing
    2. Automatic unpadding
    3. String argument to specify module output to save. E.g. --output_modules=[encoder.encoders.*]
    4. Use pytorch module forward hooks
    5. Can save encoder features or WavLM features
    6. Automatic batch dimension detection
    7. Support building ASR model from checkpoint file in inference step or from the train config in train step

Example bash script to add to asr.sh to use the python code:
if ! "${skip_data_prep}"; then
    if [ ${stage} -le 17 ] && [ ${stop_stage} -ge 17 ]; then
        _asr_train_dir="${data_feats}/${train_set}"
        _asr_valid_dir="${data_feats}/${valid_set}"

        _opts=
        if [ -n "${asr_config}" ]; then
            # To generate the config file: e.g.
            #   % python3 -m espnet2.bin.asr_train --print_config --optim adam
            _opts+="--config ${asr_config} "
        fi

        _feats_type="$(<${_asr_train_dir}/feats_type)"
        if [ "${_feats_type}" = raw ]; then
            _scp=wav.scp
            # "sound" supports "wav", "flac", etc.
            if [[ "${audio_format}" == *ark* ]]; then
                _type=kaldi_ark
            else
                _type=sound
            fi
            _fold_length="$((asr_speech_fold_length * 100))"
            _opts+="--frontend_conf fs=${fs} "
        else
            _scp=feats.scp
            _type=kaldi_ark
            _fold_length="${asr_speech_fold_length}"
            _input_size="$(<${_asr_train_dir}/feats_dim)"
            _opts+="--input_size=${_input_size} "

        fi
        if [ "${feats_normalize}" = global_mvn ]; then
            # Default normalization is utterance_mvn and changes to global_mvn
            _opts+="--normalize=global_mvn --normalize_conf stats_file=${asr_stats_dir}/train/feats_stats.npz "
        fi

        _opts+="--train_data_path_and_name_and_type ${_asr_train_dir}/${_scp},speech,${_type} "
        _opts+="--train_data_path_and_name_and_type ${_asr_train_dir}/text,text,text "
        _opts+="--train_shape_file ${asr_stats_dir}/train/speech_shape "
        _opts+="--train_shape_file ${asr_stats_dir}/train/text_shape.${token_type} "

        if ${precompute_gpu_inference}; then
            _cmd="${cuda_cmd}"
            _ngpu=1
        else
            _cmd="${decode_cmd}"
            _ngpu=0
        fi

        for dset in ${train_set}; do
            _data="${data_feats}/${dset}"
            _dir="${dumpdir}/precompute/${dset}"
            _logdir="${_dir}/logdir"
            mkdir -p "${_logdir}"

            _feats_type="$(<${_data}/feats_type)"
            if [ "${_feats_type}" = raw ]; then
                _scp=wav.scp
                if [[ "${audio_format}" == *ark* ]]; then
                    _type=kaldi_ark
                else
                    _type=sound
                fi
            else
                _scp=feats.scp
                _type=kaldi_ark
            fi

            # 1. Split the key file
            key_file=${_data}/${_scp}
            split_scps=""
            # Now only _nj=1 is implemented
            _nj=1

            for n in $(seq "${_nj}"); do
                split_scps+=" ${_logdir}/keys.${n}.scp"
            done
            # shellcheck disable=SC2086
            utils/split_scp.pl "${key_file}" ${split_scps}

            # 2. Submit decoding jobs
            log "Precomputing features... log: '${_logdir}/precompute-asr-feats.*.log'"
            rm -f "${_logdir}/*.log"
            # shellcheck disable=SC2046,SC2086
            ${_cmd} --gpu "${_ngpu}" JOB=1:"${_nj}" "${_logdir}"/precompute-asr-feats.JOB.log \
                ${python} -m utils.precompute-asr-feats \
                    --batch_size ${precompute_batch_size} \
                    --ngpu "${_ngpu}" \
                    --data_path_and_name_and_type "${_data}/${_scp},speech,${_type}" \
                    --key_file "${_logdir}"/keys.JOB.scp \
                    --asr_train_config "${asr_exp}"/config.yaml \
                    --compress true \
                    --output_dir "${_dir}/format_JOB" \
                    --use_preprocessor true \
                    --bpemodel "${bpemodel}" \
                    --token_type "${token_type}" \
                    --token_list "${token_list}" \
                    --non_linguistic_symbols "${nlsyms_txt}" \
                    --cleaner "${cleaner}" \
                    --g2p "${g2p}" \
                    --init_param ${pretrained_model} \
                    --ignore_init_mismatch ${ignore_init_mismatch} \
                    --output_modules "frontend.upstream" \
                    ${_opts} ${asr_args} || { cat $(grep -l -i error "${_logdir}"/precompute-asr-feats.*.log) ; exit 1; }
                    #--asr_model_file "${asr_exp}"/"${inference_asr_model}" \
        done
    fi

**Comment by sw005320** [comment]
Thanks!
We're now busy with the ICASSP deadline.
Also, we have one CI issue, which must be fixed.
So, we'll check your PR in a week or so.
Sorry about it...

**Comment by cyaaronk** [comment]
No problem :)

**Review by simpleoier — COMMENTED** [review]
@cyaaronk Thanks for your work!
IMO, this file is better to be put in `espnet2/bin` or `egs2/TEMPLATE/asr1/pyscripts/utils`. 
Can you give an example of how to use this feature? It would be good if this new feature is integrated in `asr1.sh`, for example, at step 3?

**Comment by cyaaronk** [comment]
@simpleoier : 
Thanks for the comment!

my current implementation can only precompute feats for asr. However, if we were to precompte other feats like speaker emb, I think we may need a separate script, and all of these should be put in a common folder. Let me know where you think is best to put it.

Also I realised many files would be affected if I modify a new model to take in the precomputed features as input, so I did not implement an example. I can try if you give me some more suggestions for the best practice.

Best regards,
Aaron

**Comment by cyaaronk** [comment]
As for an example to precompute features, please see the bash script I posted earlier

Best regards,
Aaron

**Comment by simpleoier** [comment]
Hi @cyaaronk , thanks for the updates. I'm thinking that if it is only for asr models, maybe you can put it in `egs2/TEMPLATE/asr1/pyscripts/utils`. Then you may modify the `asr.sh` in stage 3 to use some pre-trained models. In addition, after the feature extraction, you may also need to aggregate the scp files of all parallel jobs.
@ftshijt It would be better if you can give some suggestions based on your experience. This PR not only dumps ssl features, but also any layer in the model.

**Comment by ftshijt** [comment]
Hi @cyaaronk ! 

Many thanks for the contribution. We have also been concurrently adding the features (it was implemented but not published yet). Our implementation strategy depends on https://github.com/espnet/espnet/blob/master/espnet2/main_funcs/collect_stats.py#L94-L103. The benefits of using this is:
- it can be naturally adopted to all the tasks with all the frontend (e.g., fbank, s3prl-based ones, and their fusions) in ESPnet (as for all the tasks, we will have the collect stats stage)
- it has the similar options' availability with the configuration file (that is for all possible design in the frontend feature extractor, we have the accessibility with this option)

But we do find some issues with directly using this, including:
- the default setting for collect stats are for cpu jobs, it can be extend to single GPUs, but does not good for multi-gpu settings without some modification
- currently, we store each file into a npy which makes many small files (sometimes it would be fine, but for large dataset, it is not friendly, better structure might be aggregated npy files with slicing or kaldi-based ark files)
- at the time of implementation, we only utilize the single layer of SSL models so we do not need to concern about the weighted sum etc. But if to extend to multi-layers, that should be considered with more cares.

Given the time limits and other stuffs going on, we can not make time to revisit this part thoroughly (but we will soon release our basic implementation via the [new task unsupervised ASR](https://github.com/espnet/espnet/issues/4749)) So it would be definitely great if you can consider contributing on this direction. As we will probably do many cleaning and refactoring of the codebase, we did not share our implementation in the public repo yet . If you are interested, I can invite you to the private repo where we could have more discussions.

Best,
Jiatong

**Comment by ftshijt** [comment]
On the other hand, if we keep it for ASR purpose/selecte SSL only, I think it would be still a good option to follow @simpleoier 's suggestion that move the preparation into `scripts` or `pyscripts` as it is more like a supporting functions.

**Comment by cyaaronk** [comment]
@ftshijt
I am interested in contributing to your repo.
Please invite me so we can collaborate :)

Best regards,
Aaron

**Comment by ftshijt** [comment]
> cyaaronk

Many thanks for your interest. Invitation sent.
